wizwalker/extensions/scripting/deck_builder.py

```diff
@@ -189,30 +205,491 @@ def calculate_icon_position(
+        y = rectangle.y1 + xy[1]
+        return x, y
 
-    async def remove_by_predicate(self, predicate: callable, number_of_copies: Optional[int]):
-        pass
+    async def click_card(self, x: int, y: int):
+        async with self.client.mouse_handler:
+            await self.client.mouse_handler.click(x,y)
+
+
+
+    async def click_card_spell_list(self, card_num: int, number_of_copies: Optional[int]):
+        SpellList = await _maybe_get_named_window(self.client.root_window, "SpellList")
+        SpellListControl = DynamicSpellListControl(self.client.hook_handler, await SpellList.read_base_address())
+        x, y = await self.get_spell_xy(card_num, SpellListControl)
+        for _ in range(number_of_copies):
+            await self.click_card(x, y)
+
+
+    async def go_to_card_page(self, card_index):
+        SpellList = await _maybe_get_named_window(self.client.root_window, "SpellList")
+        SpellListControl = DynamicSpellListControl(self.client.hook_handler, await SpellList.read_base_address())
+        await SpellListControl.write_card_page(card_index)
 
-    async def add_by_name(self, name: str, number_of_copies: Optional[int]):
+
+
+    async def add_by_name(self, name: str, number_of_copies: Optional[int], tc = False):
         """
-        builder.add_card_by_name("unicorn", number_of_copies: int | None)
+        builder.add_card_by_name("unicorn", number_of_copies: int | None, tc: bool = False )
         -> number_of_copies = None: add max copies
         -> raises: ValueError(already at max copies)
         -> raises: ValueError(card not found)
         """
-        pass
+        card = await self.spell_list_get_card_named(name)
+
+        if number_of_copies == None:
+            number_of_copies = (await card.max_copies()) - (await card.current_copies())
+
+        if await card.max_copies() == await card.current_copies():
+            raise ValueError(f"already at max copies for {name}")
+        elif await card.max_copies() < (await card.current_copies()) + (number_of_copies):
+            raise ValueError(f"number of copies is greater than the card allows")
+
+        card_index, card_num = await self.get_card_index(name, tc) 
+        await self.go_to_card_page(card_index)
+
+        await self.click_card_spell_list(card_num, number_of_copies)
+
+
+    async def get_deck_card_index(self, card_name: str) -> int:
+        cards_names = await self.get_deck_card_names(False)
+        cards_lower = [c.lower() for c in cards_names]
+        card_placement = cards_lower.index(card_name.lower())
+        return card_placement + 1
+
+
+    async def deck_card_number_of_copies(self, card_name: str) -> int:
+        cards_names = await self.get_deck_card_names(False)
+        cards_lower = [c.lower() for c in cards_names]
+        duplicate_cards = [i for i in cards_lower if i == card_name.lower()]
+        return len(duplicate_cards)
+
+
+    async def get_item_card_index(self, card_name: str) -> list[int]:
```

Comment:
What if I got multiple of the same tc and want to use only a packed variant for example